File: js/base/async/async_connection.js
"use strict";

const EventEmitter = require('events')
const {
    NotSupported
} = require('../errors')

module.exports = class AsyncConnection extends EventEmitter {
    /*
    Events: 
        on('close')
        on('error', reason)
        on('message', data)
    */
    constructor () {
        super();
    }

    connect () {
        throw new NotSupported('not implemented method <connect>');
    }

    close () {
        throw new NotSupported('not implemented method <close>');
    }

    send(data) {
        throw new NotSupported('not implemented method <send>');
    }

    sendJson(data) {
        this.send (JSON.stringify(data));
    }
};

File: python/ccxt/base/async/async_connection.py
# -*- coding: utf-8 -*-

""" AsyncConnection """

__all__ = [
    'AsyncConnection'
]

from pyee import EventEmitter
from abc import ABC, abstractmethod
import json


class AsyncConnection (ABC, EventEmitter):
    def __init__(self):
        super(AsyncConnection, self).__init__()

    @abstractmethod
    def connect(self):
        pass

    @abstractmethod
    def close(self):
        pass

    @abstractmethod
    def send(self, data):
        pass

    def sendJson(self, data):
        self.send(json.dumps(data))
